### DISCUSSION TITLE
`docker compose build` running on action runners uses wrong path to build service images

### DISCUSSION THREAD

**Discussion by ddadaal** [community/community#61819] [Actions]
### Select Topic Area

Bug

### Body

The repo to reproduce this problem: https://github.com/ddadaal/github-actions-docker-compose-build-context-issue

This repo has  the following structure: 

```
.
├── .github
│   └── workflows
│       └── test.yaml
└── nested
    ├── docker-compose.yaml
    └── Dockerfile
```

where the `docker-compose.yaml` defines a service that uses a image built from `nested/Dockerfile`. 

We assume that `docker-compose.yaml` can only be used from the root directory of the repo, so the `docker-compose.yaml` is defined as 

```yaml
version: "3"

services:
  service1:
    image: service1-image
    build:
      context: nested
```

So the service can be started using 

```
sudo docker compose -f nested/docker-compose.yaml up -d
```

It works fine locally:

```
~/Code/github-actions-docker-compose on master                                                         at 17:51:51
❯ docker compose version
Docker Compose version v2.19.1
~/Code/github-actions-docker-compose on master                                                         at 17:48:03
❯ sudo docker compose -f nested/docker-compose.yaml build --no-cache
[+] Building 0.2s (5/5) FINISHED                                                                                   
 => [service1 internal] load .dockerignore                                                                    0.0s
 => => transferring context: 2B                                                                               0.0s
 => [service1 internal] load build definition from Dockerfile                                                 0.0s
 => => transferring dockerfile: 53B                                                                           0.0s
 => [service1 internal] load metadata for docker.io/library/alpine:3.18                                       0.0s
 => CACHED [service1 1/1] FROM docker.io/library/alpine:3.18                                                  0.0s
 => [service1] exporting to image                                                                             0.0s
 => => exporting layers                                                                                       0.0s
 => => writing image sha256:ad8e26912d6f13f9cc91c7cd7a118a4649fee37df31b009edc3127a03995ccbd                  0.0s
 => => naming to docker.io/library/service1-image                                                             0.0s

~/Code/github-actions-docker-compose on master                                                         at 17:48:12
❯ sudo docker compose -f nested/docker-compose.yaml up -d           
[+] Running 1/1
 ✔ Container nested-service1-1  Started 
```

Then, the repo defines a workflow that calls exactly the command above (`docker compose -f nested/docker-compose.yaml up -d`), but it failed to run ( https://github.com/ddadaal/github-actions-docker-compose-build-context-issue/actions/runs/5667130914/job/15355241715 ). 

The error is `unable to prepare context: path "/home/runner/work/github-actions-docker-compose-build-context-issue/github-actions-docker-compose-build-context-issue/nested/nested" not found`. It seems the path build path is wrong. It should be `nested` but actually is`nested/nested`.

This bug only occurs starting from 2pm (+8 UTC) of 2023/7/26. It works fine before.

**Comment by Beargruug** [comment]
The error message says, "unable to prepare context: path '/home/runner/work/github-actions-docker-compose-build-context-issue/github-actions-docker-compose-build-context-issue/nested/nested' not found." It seems like the build path is incorrect. Instead of using "nested" as the build context, it's trying to use "nested/nested," which doesn't exist and causes the failure.

To fix this, we need to adjust the GitHub Actions workflow file. Specifically, we need to specify the correct build context in the workflow.

Here's what we can do:

1. Go to the ".github/workflows" directory in the repository, and find the "test.yaml" file.

2. Inside the "test.yaml" file, we'll make a few changes to ensure the build context is set correctly.

3. We'll add a "working-directory" field under the "Build and Run Docker Compose" step and set it to "./nested". This tells GitHub Actions to run the Docker Compose commands from the "nested" directory.

4. Once we've made the changes, we can commit the updated "test.yaml" file back to the repository.

The updated part of the "test.yaml" file should look like this:

```
# ... other steps ...

  - name: Build and Run Docker Compose
    working-directory: ./nested
    run: |
      sudo docker-compose build --no-cache
      sudo docker-compose up -d

# ... other steps ...
```
By specifying the correct working directory, the build context will be set to "nested," and the Docker Compose setup should run successfully without the "nested/nested" path issue.

I hope this will help you!

**Reply by ddadaal** [reply]
This is what happens if `docker compose` is run under `nested` directory instead of root directory: 

```
❯ sudo docker compose build   
[+] Building 0.0s (0/0)                                                                                            
unable to prepare context: path "nested" not found
```

The configuration file specifies that the context which the image is built on is `nested` **relative to cwd**, so the configuration file can only be used under root directory so that the final path is correctly `<root>/nested`.

**Comment by ddadaal** [comment]
OK I figured it out. The [docker compose docs](https://docs.docker.com/compose/compose-file/build/#context) states that the build context is relative to the compose file, not cwd. It is still strange though why my previous config worked.

**Reply by dimitarvp** [reply]
What did you end up doing exactly? I am having the same issue as your original one and I am not sure what to do. The docs you linked to didn't enlighten me either. Could you show a diff, please?
